File: src/utils/gemini-client.ts
import requests from 'node:http';
import urllib from 'node:url';

const URL = "https://gemini.google.com/_/BardChatUi/data/assistant.lamda.BardFrontendService/StreamGenerate";

const HEADERS = {
  "accept": "*/*",
  "content-type": "application/x-www-form-urlencoded;charset=UTF-8",
  "x-same-domain": "1",
  "cookie": "" // Add your cookie here if needed
};

function buildPayload(prompt: string): string {
  const inner = [
    [prompt, 0, null, null, null, null, 0],
    ["en-US"],
    ["", "", "", null, null, null, null, null, null, ""],
    "", "", null, [0], 1, null, null, 1, 0,
    null, null, null, null, null, [[0]], 0
  ];

  const outer = [null, JSON.stringify(inner)];

  return new URLSearchParams({
    "f.req": JSON.stringify(outer)
  }).toString() + "&";
}

function parseResponse(text: string): string {
  let cleanedText = text.replace(")]}'", "");
  let best = "";

  for (const line of cleanedText.split('\n')) {
    if (!line.includes("wrb.fr")) {
      continue;
    }

    try {
      const data = JSON.parse(line);
      let entries = [];
      
      if (Array.isArray(data)) {
        if (data[0] === "wrb.fr") {
          entries = [data];
        } else {
          entries = data.filter((i: any) => Array.isArray(i) && i[0] === "wrb.fr");
        }
      }

      for (const entry of entries) {
        try {
          const inner = JSON.parse(entry[2]);

          if (Array.isArray(inner) && Array.isArray(inner[4])) {
            for (const c of inner[4]) {
              if (Array.isArray(c) && Array.isArray(c[1])) {
                const txt = c[1]
                  .filter((t: any) => typeof t === 'string')
                  .join('');
                if (txt.length > best.length) {
                  best = txt;
                }
              }
            }
          }
        } catch {}
      }
    } catch {}
  }

  return best.trim();
}

export async function askGemini(prompt: string): Promise<string> {
  const payload = buildPayload(prompt);

  try {
    const response = await fetch(URL, {
      method: 'POST',
      headers: HEADERS,
      body: payload
    });

    if (response.status !== 200) {
      return `[ERROR ${response.status}]`;
    }

    const text = await response.text();
    return parseResponse(text) || "[No response]";
  } catch (error) {
    return `[Error: ${error instanceof Error ? error.message : 'Unknown error'}]`;
  }
}

File: src/utils/gemini-client.py
import requests
import json
import urllib.parse

URL = "https://gemini.google.com/_/BardChatUi/data/assistant.lamda.BardFrontendService/StreamGenerate"

HEADERS = {
    "accept": "*/*",
    "content-type": "application/x-www-form-urlencoded;charset=UTF-8",
    "x-same-domain": "1",
    "cookie": ""  # waywaaa
}


def build_payload(prompt):
    inner = [
        [prompt, 0, None, None, None, None, 0],
        ["en-US"],
        ["", "", "", None, None, None, None, None, None, ""],
        "", "", None, [0], 1, None, None, 1, 0,
        None, None, None, None, None, [[0]], 0
    ]

    outer = [None, json.dumps(inner)]

    return urllib.parse.urlencode({
        "f.req": json.dumps(outer)
    }) + "&"


def parse_response(text):
    text = text.replace(")]}'", "")
    best = ""

    for line in text.splitlines():
        if "wrb.fr" not in line:
            continue

        try:
            data = json.loads(line)
        except:
            continue

        entries = []
        if isinstance(data, list):
            if data[0] == "wrb.fr":
                entries = [data]
            else:
                entries = [i for i in data if isinstance(i, list) and i[0] == "wrb.fr"]

        for entry in entries:
            try:
                inner = json.loads(entry[2])

                if isinstance(inner, list) and isinstance(inner[4], list):
                    for c in inner[4]:
                        if isinstance(c, list) and isinstance(c[1], list):
                            txt = "".join([t for t in c[1] if isinstance(t, str)])
                            if len(txt) > len(best):
                                best = txt
            except:
                continue

    return best.strip()


def ask(prompt):
    payload = build_payload(prompt)

    res = requests.post(URL, headers=HEADERS, data=payload)

    if res.status_code != 200:
        return f"[ERROR {res.status_code}]"

    return parse_response(res.text) or "[No response]"


# ===== CHAT LOOP =====
print("⚠️ Unofficial Gemini Chatbot (type 'exit' to quit)\n")

while True:
    user_input = input("You: ")

    if user_input.lower() in ["exit", "quit"]:
        break

    reply = ask(user_input)
    print("Bot:", reply)
